add UX documentation (#61)
qiskit-community/spank-plugins: docs/ux.md

The source code does not have to take care resp. is not involved in resource reservation handling (that is done when slurm jobs are assigned QPU resources and start running, if applicable on the backend) or execution modes like sessions (these are automatically in place while the job is running, if applicable on the backend).
This makes the source code more portable between similar QPU resource types through different backend access methods (such as IBM's Direct Access API and IBM's Qiskit Runtime service through IBM Quantum Platform).
All backend types (such as IBM's Direct Access API, IBM's Qiskit Runtime service, or Pasqal's backends) follow these principles.

## Connecting physical resources to slurm resoures and how to use them

Note the exact syntax is subject to change -- this is a sketch of the UX at this time.

### HPC admin scope

HPC administrators configure, what physical resources can be provided to slurm jobs.

```
# slurm quauntum plugin configuration

name=da-local-backend                            \
url=https://da-endpoint.my-local.domain/       \
s3_url=https://s3.my-local.domain/...          \
s3_accesstoken=4828d9bea...

> # crn, credential and s3 fields can be set for all users set or overridden by users individually

Reviewer: I might be confused, but I think this is related to UX discussion.

It seems HPC admin will create/set api key and S3 bucket credential to be shared with slurm users for their job executions. Is this correct? 

I don't think sharing API keys is recommended, at least in IBM Quantum cloud. I believe that each user is required to create and work with their own apikey when accessing the backend. If we use shared api key for job execution, I think we need to consider another mechanism to check if each user has access to the quantum resource, such as checking the denied party list?

Also for the S3 bucket, I think it would be better to allow users to access the results/logs in S3 bucket directly. In this case, the user would specify the S3 credential.

Author: Interesting discussion which path (personal tokens vs one cluster-wide access) is desired by all stakeholders. It sounds like a pragmatic way forward to go with personal tokens for now, thank you for the comment. I will update the PR!

S3 bucket access is hopefully encapsulated by the primitive classes? Having users define S3 credentials does not sound convenient. Let's discuss, whether we want to offer both inline and S3 data transfers etc.

HPC user experience, HPC developer experience and usage patterns
================================================================

## Content

- [Principles](#principles)
- [Connecting physical resources to slurm resoures and how to use them](#connecting-physical-resources-to-slurm-resoures-and-how-to-use-them)
  - [HPC admin scope](#hpc-admin-scope)
  - [HPC user scope](#hpc-user-scope)
  - [HPC application scope](#hpc-application-scope)
  - [Backend specifics](#backend-specifics)
    - [IBM Direct Access API](#ibm-direct-access-api)
    - [Qiskit Runtime Service](#qiskit-runtime-service)
- [Examples](#examples)
  - [Running jobs with dependencies](#running-jobs-with-dependencies)
  - [Running a job with several slurm QPU resources](#running-a-job-with-several-slurm-qpu-resources)
  - [Running primitives directly](#running-primitives-directly)
  - [Other workflow tools](#other-workflow-tools)

See [Overview](./overview.md) for a glossary of terms.

## Principles

Slurm QPU resource definitions determine what physical resources can be used by slurm jobs.
User source code should be agnostic to specific backend instances and even backend types as far as possible.

> This keeps source code portable while the QPU seletion criteria are part of the resource definition (which is considered configuration as opposed to source code).

Reviewer 2: ```suggestion
This keeps source code portable while the QPU selection criteria are part of the resource definition (which is considered configuration as opposed to source code).
```

The source code does not have to take care resp. is not involved in resource reservation handling (that is done when slurm jobs are assigned QPU resources and start running, if applicable on the backend) or execution modes like sessions (these are automatically in place while the job is running, if applicable on the backend).
This makes the source code more portable between similar QPU resource types through different backend access methods (such as IBM's Direct Access API and IBM's Qiskit Runtime service through IBM Quantum Platform).
All backend types (such as IBM's Direct Access API, IBM's Qiskit Runtime service, or Pasqal's backends) follow these principles.

## Connecting physical resources to slurm resoures and how to use them

Note the exact syntax is subject to change -- this is a sketch of the UX at this time.

### HPC admin scope

HPC administrators configure, what physical resources can be provided to slurm jobs.

> Note sensitivethe config could contain sensitive information and m

Reviewer 2: ```suggestion
Note that the config could contain sensitive information and m
```
Not sure what @utzb might have wanted to continue writing at the end here..

```

> # slurm quauntum plugin configuration

Reviewer 2: ```suggestion
# slurm quantum plugin configuration
```

# DA backend
name=da-local-backend                                                    \
url=https://da-endpoint.my-local.domain/                                 \
da_crn=crn:v1:bluemix:public:quantum-computing:us-east:a/43aac17...      \
da_apikey_file=/root/da_apikey                                           \
s3_url=https://s3.my-local.domain/...                                    \
s3_accesstoken_file=/root/s3_accesstoken

# QRS backends
name=ibm_fez                                                             \
type=qiskit-runtime-service-ibmcloud

name=ibm_marrakesh                                                       \

> type=qiskit-runtime-service-ibmcloud

Reviewer 2: I might just not have seen this config format before, but where does this get parsed and how will that system handle repeating keys (e.g. the `name` key being re-used without "sections")?

Author: the format aligns to what seems typical of gres.conf content. Adding a comment.

```

See the specific sections of the backend type for details on the parameters.

In `slurm.conf`, qpu generic resources can be assigned to some or all nodes for usage:
```
...
GresTypes=qpu,name
NodeName=node[1-5000] Gres=qpu,name:ibm_fez
...
```

### HPC user scope

HPC users submit jobs using QPU resources that are tied to slurm QPU resources.
The name attribute references what the HPC administrator has defined.
Mid-term, backend selection can be based on criteria other than a predefined name which refers to a spefific backend (e.g. by capacity and error rate qualifiers which help downselect between the defined set of backends)

Slurm qpu resources given an identifier (in this example: *my_qpu_resource*) that can be referenced by applications e.g. to distinguish between several assigned resources.

```shell
#SBATCH --time=100
#SBATCH --output=<LOGS_PATH>
#SBATCH --qpu=my_qpu_resource

> #SBATCH --qpu-name=ibm_fez

Author: right -- I will add this example to the text!

#SBATCH --... # other options

srun ...
```

There might be additional environment variables required, depending on the backend type.

### HPC application scope

HPC applications refer to the slurm QPU resources assigned to the slurm job.

(Please check whether this flow clicks. We want to reference the slurm resource name, in case several resources are defined; also we ideally avoid specifying whether it's a DirectAccess or QiskitRuntimeService (Pasqal code is probably not very portable to other types at this time); also it seems we wanted to avoid a backend object and go with target instead)

```python
from qrmi import IBMDirectAccessSamplerV2 as SamplerV2

# Generate transpiler target from backend configuration & properties

> target = get_target(slurm_resource_name=“my_qpu_resource”)

Reviewer 2: Building on top of my previous comment, this still "couples" this application code to how the user configures `sbatch`. Can we not do something like:
```python
target = get_target(slurm_resource_name=os.environ["SLURM_QPU_NAME"])
```
where the `SLURM_QPU_NAME` environment variable gets automatically exported by our plugin?

See e.g. [sbatch's output env variables](https://slurm.schedmd.com/sbatch.html#SECTION_OUTPUT-ENVIRONMENT-VARIABLES).

---

More specifically, in the context of a spank plugin we might obtain `SPANK_...` (e.g. `SPANK_QPU_NAME`) according to https://slurm.schedmd.com/spank.html

> Functions are also available from within the SPANK plugins to establish environment variables to be exported to the Slurm PrologSlurmctld, Prolog, Epilog and EpilogSlurmctld programs (the so-called job control environment). The name of environment variables established by these calls will be prepended with the string SPANK_ in order to avoid any security implications of arbitrary environment variable control. (After all, the job control scripts do run as root or the Slurm user.).

Author: let me try to propose something

# The circuit and observable need to be transformed to only use instructions
# supported by the QPU (referred to as instruction set architecture (ISA) circuits).
# We'll use the transpiler to do this.
pm = generate_preset_pass_manager(
    optimization_level=1,
    target=target,
)

sampler = SamplerV2(target=target)
```

### Backend specifics
#### IBM Direct Access API
##### HPC admin scope
Configuration of Direct Access API backends (HPC admin scope) includes endpoints and credentials to the Direct Access endpoint, authentication services as well as the S3 endpoint.
Specifically, this includes:

* IBM Cloud API key for creating bearer tokens
* endpoint of Direct Access API
* S3 bucket and access details

Access credentials should not visible to HPC users or other non-privileged users on the system.
Therefore, sensitive data is put in separate files which can be access protected accordingly.

> Note that all users share the same access and hence need appropriate user vetting before getting access to quantum resources.

Author: adding comments for clarification

This makes the source code more portable between similar QPU resource types through different backend access methods (such as IBM's Direct Access API and IBM's Qiskit Runtime service through IBM Quantum Platform).
All backend types (such as IBM's Direct Access API, IBM's Qiskit Runtime service, or Pasqal's backends) follow these principles.

## Connecting physical resources to slurm resoures and how to use them

Note the exact syntax is subject to change -- this is a sketch of the UX at this time.

### HPC admin scope

HPC administrators configure, what physical resources can be provided to slurm jobs.
The config could contain sensitive information such as credentials.
Moving such items into files and referencing these can make it easier to protect the information and rotate credentials, if HPC admins prefer to not add credentials right in the configuration file.

Here is an example configuration (the format aligns to the typical content of gres.conf -- one line per resource):

```
# slurm quantum plugin configuration

> # DA backend

Reviewer 2: Just noting that this backend does not have a `type` while the QRS backends below do. Is there an implied default here?

Author: good catch!

name=da-local-backend                                                    \
url=https://da-endpoint.my-local.domain/                                 \
da_crn=crn:v1:bluemix:public:quantum-computing:us-east:a/43aac17...      \
da_apikey_file=/root/da_apikey                                           \
s3_url=https://s3.my-local.domain/...                                    \
s3_accesstoken_file=/root/s3_accesstoken

# QRS backends
name=ibm_fez                                                             \
type=qiskit-runtime-service-ibmcloud

name=ibm_marrakesh                                                       \
type=qiskit-runtime-service-ibmcloud
```

See the specific sections of the backend type for details on the parameters.

In `slurm.conf`, qpu generic resources can be assigned to some or all nodes for usage:
```
...
GresTypes=qpu,name
NodeName=node[1-5000] Gres=qpu,name:ibm_fez
...
```

### HPC user scope

HPC users submit jobs using QPU resources that are tied to slurm QPU resources.
The name attribute references what the HPC administrator has defined.

> Mid-term, backend selection can be based on criteria other than a predefined name which refers to a spefific backend (e.g. by capacity and error rate qualifiers which help downselect between the defined set of backends).

Reviewer 2: ```suggestion
Mid-term, backend selection can be based on criteria other than a predefined name which refers to a specific backend (e.g. by capacity and error rate qualifiers which help downselect between the defined set of backends).
```

There might be additional environment variables required, depending on the backend type.

#### Single QPU resources

When a single QPU resource is used, SBATCH parameters will point to the QPU resource assigned to the application.
Environment variables provided through the plugin can provide more details to the application (e.g. `SLURM_QPU_NAME` indicating which QPU resource is used eventually).

```shell
#SBATCH --time=100
#SBATCH --output=<LOGS_PATH>
#SBATCH --qpu=ibm_fez
#SBATCH --... # other options

srun ...
```

#### Multiple QPU resources

Slurm qpu resources are given an identifier (in this example: *my_qpu_resource*) that can be referenced by applications e.g. to distinguish between several assigned resources.

> In this example, *my_qpu_resource* is the reference for usage of the quantum resource in the application.

Reviewer 2: This implies that the single vs multiple QPU case is implemented differently, the former accessing runtime information such as the QPU name from an environment variable while the latter should write these (presumably fixed) aliases. Or am I misunderstanding this?

What stops us from always using environment variables? At that point, we don't necessarily need this additional level of alias/indirection. It would also be more streamlined.

For example:
```
SLURM_QPU_COUNT = #integer : number of allocated QPU resources
SLURM_QPU_NAMES = #list[str] : names of the QPU resources of length SLURM_QPU_COUNT
```
In this way, a program could even check how many QPUs have been allocated at runtime, and it will always know where to find _all_ their names (and from this can query their properties).
One could even infer the number of QPU resources from the length of `SLURM_QPU_NAMES` to avoid one more variable if necessary.

Author: having Linux backgroud, I like colon separated environment variables, but that's really cool!

(Please check whether this flow clicks. We want to reference the slurm resource name, in case several resources are defined; also we ideally avoid specifying whether it's a DirectAccess or QiskitRuntimeService (Pasqal code is probably not very portable to other types at this time); also it seems we wanted to avoid a backend object and go with target instead)

#### Single QPU resources

For single QPU scenarios, the QPU will be identified and used automatically.

```python
from qrmi import IBMDirectAccessSamplerV2 as SamplerV2

# Generate transpiler target from backend configuration & properties
target = get_target()
pm = generate_preset_pass_manager(
    optimization_level=1,
    target=target,
)

sampler = SamplerV2(target=target)
```

#### Multiple QPU resources

A named identifier represents the QPU resource and can be used in the source code.

```python
from qrmi import IBMDirectAccessSamplerV2
from qrmi import IBMQiskitRuntimeServiceSamplerV2

# Generate transpiler target from backend configuration & properties
target1 = get_target(slurm_resource_name=“my_qpu_resource1”)
target2 = get_target(slurm_resource_name=“my_qpu_resource2”)

> target2 = get_target(slurm_resource_name=“my_qpu_resource3”)

Reviewer 2: My suggestion above would result in this being the following:
```python
num_targets = os.environ["SLURM_QPU_COUNT"]
targets = [get_target(id=i) for i in range(num_targets)]
```
where
```python
def get_target(index: int = 0) -> Target:
    target_names = os.environ["SLURM_QPU_NAMES"].split(";")  # assuming we have a ";"-separated list of names
    target_name = target_names[index]
    # obtain all target properties and whatever we return here
    return Target.from_name(target_name)
```

Reviewer 2: If we do not like the `SLURM_QPU_NAMES` we could also have separate env variables for each name be defined:
```
SLURM_QPU_NAME0 = #str
SLURM_QPU_NAME1 = #str
...  # continues up to `SLURM_QPU_NAME{SLURM_QPU_COUNT}`
```
In this case we would need a `SLURM_QPU_COUNT` variable which determines the number of variables defined.

# The circuit and observable need to be transformed to only use instructions
# supported by the QPU (referred to as instruction set architecture (ISA) circuits).
# We'll use the transpiler to do this.
pm1 = generate_preset_pass_manager(
    optimization_level=1,
    target=target1,
)

pm2 = generate_preset_pass_manager(
    optimization_level=1,
    target=target2,
)

pm3 = generate_preset_pass_manager(
    optimization_level=1,
    target=target3,
)

sampler1 = IBMDirectAccessSamplerV2(target=target1)
sampler2 = IBMQiskitRuntimeServiceSamplerV2(target=target2)
sampler3 = IBMQiskitRuntimeServiceSamplerV2(target=target3)
```

### Backend specifics
#### IBM Direct Access API
##### HPC admin scope
Configuration of Direct Access API backends (HPC admin scope) includes endpoints and credentials to the Direct Access endpoint, authentication services as well as the S3 endpoint.
Specifically, this includes:

* IBM Cloud API key for creating bearer tokens
* endpoint of Direct Access API
* S3 bucket and access details

> Access credentials should not visible to HPC users or other non-privileged users on the system.

Reviewer 2: ```suggestion
Access credentials should not be visible to HPC users or other non-privileged users on the system.
```
